radix-engine/src/engine/modules/execution_trace.rs

```diff
@@ -18,41 +18,89 @@ pub struct ExecutionTraceReceipt {
 }
 
 #[derive(Debug)]
-pub struct ExecutionTrace {
-    /// Stores resource changes that resulted from vault's put/take operations.
-    pub resource_changes: HashMap<ComponentId, HashMap<VaultId, (ResourceAddress, Decimal)>>,
-
-    /// Stores component IDs associated with vaults that have been used to lock a fee.
-    /// This, together with a FeeSummary, is later used to create ResourceChange entries
-    /// for fee payments (incl. any refunds back to the vault).
-    pub fee_vaults_components: HashMap<VaultId, ComponentId>,
+pub enum VaultOp {
+    Create(Decimal), // TODO: add trace of vault creation
+    Put(Decimal),    // TODO: add non-fungible support
+    Take(Decimal),
+    LockFee,
+    LockContingentFee,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, TypeId)]
+pub enum ExecutionTraceError {}
+
+#[derive(Debug)]
+pub struct ExecutionTrace {}
+
+impl<R: FeeReserve> Module<R> for ExecutionTrace {
+    fn pre_sys_call(
+        &mut self,
+        track: &mut Track<R>,
+        call_frames: &mut Vec<CallFrame>,
+        input: SysCallInput,
+    ) -> Result<(), ModuleError> {
+        Self::trace_invoke_method(track, call_frames, input);
+        Ok(())
+    }
+
+    fn post_sys_call(
+        &mut self,
+        _track: &mut Track<R>,
+        _call_frames: &mut Vec<CallFrame>,
+        _output: SysCallOutput,
+    ) -> Result<(), ModuleError> {
+        Ok(())
+    }
+
+    fn on_wasm_instantiation(
+        &mut self,
+        _track: &mut Track<R>,
+        _call_frames: &mut Vec<CallFrame>,
+        _code: &[u8],
+    ) -> Result<(), ModuleError> {
+        Ok(())
+    }
+
+    fn on_wasm_costing(
+        &mut self,
+        _track: &mut Track<R>,
+        _call_frames: &mut Vec<CallFrame>,
+        _units: u32,
+    ) -> Result<(), ModuleError> {
+        Ok(())
+    }
+
+    fn on_lock_fee(
+        &mut self,
+        _track: &mut Track<R>,
+        _call_frames: &mut Vec<CallFrame>,
+        _vault_id: VaultId,
+        fee: Resource,
+        _contingent: bool,
+    ) -> Result<Resource, ModuleError> {
+        Ok(fee)
+    }
 }
 
 impl ExecutionTrace {
```

Comment:
now that it's a full-fledged module, should we rename it to ExecutionTraceModule?

Reply:
Yes, will do.